python/ray/air/execution/_internal/event_manager.py

```diff
@@ -117,12 +122,78 @@ class RayEventManager:
     def __init__(self, resource_manager: ResourceManager):
         self._resource_manager: ResourceManager = resource_manager
 
-        raise NotImplementedError
+        # ---
```

Comment:
I've tried to add extensive inline documentation here to make this easy to understand.

The main crux is to understand the `wait()` method. The reason why it is a bit more complex is that we have to deal with two different resource backends: One that uses futures (placement groups) and one that doesn't (budget based).

For budget-based, we have to try starting/stopping actors at different times and potentially with our own timeout mechanism as we can't await resources becoming ready.

For placement group based, we can actually await futures - in that case, we don't need our own special logic.

I tried to make the logic very clear with the in-line comments, please let me know if there is anything left unclear.

Reply:
wait, is this EventManager or ActorManager?
at a very high level, EventManager probably shouldn't manage actor life cycles.
it should only detect actor and future state changes and fire events accordingly?